cadcdata/cadcdata/core.py

```diff
@@ -0,0 +1,418 @@
+                        unicode_literals)
+
+import argparse
+import imp
+import logging
+import os
+import os.path
+import sys
+from StringIO import StringIO
+from datetime import datetime
+
+from cadcutils import net, util, exceptions
+from caom2.obs_reader_writer import ObservationReader, ObservationWriter
+from caom2.version import version as caom2_version
+from six.moves.urllib.parse import urlparse
+import zlib
+
+# from . import version as caom2repo_version
+from cadcdata import version
+
+__all__ = ['CadcDataClient']
+
+# TODO replace with SERVICE_URI when server supports it
+SERVICE_URL = 'www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/'
+# IVOA dateformat
+DATE_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
+DEFAULT_RESOURCE_ID = 'ivo://cadc.nrc.ca/data'
+ARCHIVE_STREAM_HTTP_HEADER = 'X-CADC-Stream'
+APP_NAME = 'cadc-data'
+
+READ_BLOCK_SIZE=8*1024
+
+logger = logging.getLogger(APP_NAME)
+
+class CadcDataClient:
+
+    """Class to access CADC archival data."""
+
+    logger = logging.getLogger('CadcDataClient')
+
+    def __init__(self, subject, resource_id=DEFAULT_RESOURCE_ID, host=None):
+        """
+        Instance of a CAOM2RepoClient
+        :param subject: the subject(user) performing the action
+        :type subject: Subject
+        :param resource_id: The identifier of the service resource (e.g 'ivo://cadc.nrc.ca/data')
+        :param host: Host server for the caom2repo service
+        """
+
+        self.resource_id = resource_id
+
+        # TODO This is just a temporary hack to be replaced with proper registry lookup functionaliy
+        resource_url = urlparse(resource_id)
+        self.host = host
+
+        agent = "cadc-data-client/{}".format(version.version)
+
+        self._repo_client = net.BaseWsClient(resource_id, subject,
+                                             agent, retry=True, host=self.host)
+        self.logger.info('Service URL: {}'.format(self._repo_client.base_url))
+
+
+    def get_file(self, archive, file_id, file=None, decompress=False,
+                 cutout=None, fhead=False, wcs=False, process_bytes=None):
+        """
+        Get a file from an archive. The entire file is delivered unless the cutout argument
+         is present specifying a cutout to extract from the file.
+        :param archive: name of the archive containing the file
+        :param file_id: the ID of the file to retrieve
+        :param file: file to save data to (file or file_name). If None, it uses the response
+        content disposition to determine the name of the destination file
+        :param cutout: perform cutout operation on the file before the result is delivered
+        :param wcs: True if the wcs is to be included with the file
+        :param process_bytes: function to be applied to the received bytes
+        :return: the data stream object
+        """
+        assert archive is not None
+        assert file_id is not None
+        resource = '/{}/{}'.format(archive, file_id)
+        params = {}
+        if fhead:
```

Comment:
I agree this looks clunky but that's only because it just happens that the names of the function arguments match the HTTP attributes of the service. kwargs is a great mechanism to transparently pass attributes between two layers, in this case between the client code and the data webservice. While the code would look better, it would essentially expose the implementation of our data Web service which I don't think it's a good idea especially since it's not even a standard. What happens if on CC the implementation changes or we later decide to change the names of the HTTP attributes? Or we cannot do head on the server and have to implement it in the get_file function through a seek in a file or something? All the clients would then be exposed.